Adapt custom serving runtimes to KServe

DaoDaoNoCode opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
Closes #1912 

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
The PR did the following things:
1. Add a required select dropdown when adding/editing custom serving runtimes, you can not click submit without selecting a value

<img width="1727" alt="Screenshot 2023-10-16 at 3 28 19 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/37624318/55387dc6-d3e3-4006-9c1b-e562cb3cc8ec">

2. Add a column to show the supported platform of the custom serving runtimes, and in the header show the available platforms based on the dashboard config

<img width="1727" alt="Screenshot 2023-10-18 at 5 19 04 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/37624318/ef6f7cd9-b2e6-4bdf-91f6-2aa18b4a5370">

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
To test the custom serving runtime template labels:
1. Go to the custom serving runtimes page
2. Make sure the OOTB serving runtimes have both labels shown because they don't have the `opendatahub.io/modelServingSupport` annotation
3. Try to duplicate one of the serving runtimes, make sure you cannot submit without selecting a platform
4. Select a platform and create one
5. Make sure the label is correct
6. Try to edit the serving runtime again, choose a different option, and submit
7. Make sure the label is changed too

To test the header labels of the available platforms:
1. Check the dashboard config, and try to change the `disableKServe` flag and `disableModelMesh` flag
2. Every time you make the change, wait for 2 minutes until the frontend fetches the latest data
3. If you set `disableKServe` to `true`, you shouldn't see the `Single model serving enabled` label. If you set `disableModelMesh` to `true`, you shouldn't see the `Multi-model serving enabled` label. On the other hand, setting these values to `false` should make the labels visible

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
Add the storybook test for the custom serving runtimes page and check the labels, more tests to be added along with the feature development.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

DaoDaoNoCode commented:
@vconzola Can you check the screenshots when you are available? Thanks!

lucferbux commented:
@DaoDaoNoCode first of all, I've approved https://github.com/opendatahub-io/odh-dashboard/pull/1936 can you then rebase it and check the integration is working?

lucferbux commented:
I'm gonna update something of the requirements, just to do the follow up

lucferbux commented:
Ok, @DaoDaoNoCode I've created https://github.com/opendatahub-io/odh-dashboard/issues/1973 to do a follow up, let's just get this and then get the other section so we are unblocked asap.

lucferbux commented on frontend/src/pages/modelServing/customServingRuntimes/CustomServingRuntimeAddTemplate.tsx:
```diff
@@ -101,40 +123,47 @@ const CustomServingRuntimeAddTemplate: React.FC<CustomServingRuntimeAddTemplateP
-            >
-              {error.message}
-            </Alert>
+            <CustomServingRuntimePlatformsSelector
```
Mmmmmm I'm trying to follow the logic, but somehow when clonning the Serving Runtime it doesn't update the platform selected

lucferbux commented on frontend/src/__tests__/integration/pages/customServingRuntimes/CustomServingRuntimes.stories.tsx:
```diff
@@ -0,0 +1,67 @@
+                mockServingRuntimeTemplateK8sResource({
+                  name: 'template-2',
+                  displayName: 'Test template',
+                  platforms: [ServingRuntimePlatform.MULTI],
```
Add one example more with no annotations to cover the state of a template not having platforms.

lucferbux commented on frontend/src/pages/modelServing/customServingRuntimes/CustomServingRuntimeListView.tsx:
```diff
@@ -24,6 +24,8 @@ const CustomServingRuntimeListView: React.FC = () => {
 
   const [deleteTemplate, setDeleteTemplate] = React.useState<TemplateKind>();
+
+  // TODO: add search functions
```
Don't add the TODO here, let's just control that with tickets, im gonna start a conversation to track this cause im not sure if we want that.

lucferbux reviewed (changes requested):
Just small nits, i think you can get this in no time

DaoDaoNoCode commented:
I updated this PR so #1973 will be also implemented here, I updated the screenshot in the description. Also, I addressed all the comments above, could you check it again? @lucferbux
